mqtt/mqtt-bridge/src/persist/loader.rs

```diff
@@ -4,32 +4,34 @@ use futures_util::stream::Stream;
 use mqtt3::proto::Publication;
 use parking_lot::{Mutex, MutexGuard};
 
-use crate::persist::{memory::waking_map::WakingMap, Key};
+use crate::persist::{waking_state::StreamWakeableState, Key};
 
 /// Message loader used to extract elements from bridge persistence
+///
 /// This component is responsible for message extraction from the persistence
 /// It works by grabbing a snapshot of the most important messages from the persistence
 /// Then, will return these elements in order
+///
 /// When the batch is exhausted it will grab a new batch
-pub struct InMemoryMessageLoader {
-    state: Arc<Mutex<WakingMap>>,
+pub struct MessageLoader<S: StreamWakeableState> {
+    state: Arc<Mutex<S>>,
     batch: IntoIter<(Key, Publication)>,
     batch_size: usize,
 }
 
-impl InMemoryMessageLoader {
-    pub async fn new(state: Arc<Mutex<WakingMap>>, batch_size: usize) -> Self {
+impl<S: StreamWakeableState> MessageLoader<S> {
+    pub async fn new(state: Arc<Mutex<S>>, batch_size: usize) -> Self {
         let batch: IntoIter<(Key, Publication)> = Vec::new().into_iter();
```

Comment:
this IntoIter looks a bit weird. what you could do is to use VecDeque. and instead of calling `self.batch.next()` you could do `self.batch.pop_back()`